### ISSUE TITLE
Array iterators do not have access to context

### ISSUE THREAD

**Issue by NeilMacMullen** [scriban/scriban#322]
Code such as 

```
func call_array(p)
  ret  [1,2,3] | array.filter @(do;ret p;end)
end

#This fails: parameter is not found at line 14 when calling into array iterator
call_array true
```

will return an error (when StrictVariables is true), claiming that` error : The variable or function `p` was not found`

**Comment by NeilMacMullen** [comment]
@xoofx   The reason for this appears to be that the internal implementations EachInternal and FilterInternal (which copied the pattern) *yield* an IEnumerable which is only evaluated later after the TemplateContext has been modified on its way out of the call stack.  

The easy fix is to modify the ApplyFunction to force the Enumerable to be evaluated before passing it to the ScriptRange constructor (ArrayFunctions.cs:710) and I have a potential PR for this.  However  I assume the deferred evaluation was deliberate for performance reasons?  If so maybe the better fix is to figure out why the TemplateContext is being modified after the iterator?

**Comment by xoofx** [comment]
> The easy fix is to modify the ApplyFunction to force the Enumerable to be evaluated before passing it to the ScriptRange constructor (ArrayFunctions.cs:710) and I have a potential PR for this. However I assume the deferred evaluation was deliberate for performance reasons? 

It's because you can have ranges `1..1000000` and they are not evaluated today as a list but as an IEnumerable.

> If so maybe the better fix is to figure out why the TemplateContext is being modified after the iterator?

Hm, that's where it's super troublesome. Scriban doesn't capture the context of variables (e.g as JavaScript is doing for example). That would be tricky to bring (and potentially a performance nightmare...)...
Not sure what to do with that... don't have the head enough in Scriban these days to fix that easily... 😅

**Comment by xoofx** [comment]
> Not sure what to do with that... don't have the head enough in Scriban these days to fix that easily... 😅

Maybe a potential easy fix is to convert a `ScriptRange` to a `ScriptArray` when a ret is issued so that it forces the capture?

**Comment by NeilMacMullen** [comment]
Yes - I was thinking of something similar - forcing the ScriptRange to perform the eval before the context got lost.  I will see if I can find a suitable place to do so....

**Comment by NeilMacMullen** [comment]
(FWIW - I understand your point about ranges but closing over local variables for array iterators would be super useful for me and a nice language feature I think!)

**Comment by xoofx** [comment]
Fixed by #324
